Add Terraform & HCL syntax highlighting

caius opened the pull request:
Terraform and HCL are almost the same language, but not quite so proposing them as separate languages within Zed. (Terraform is an extension of HCL, with a different formatter.)

This is just adding the language definition, parsing and highlighting functionality, not any LSP or formatting beyond that for either language.

I've taken a bunch of inspiration from Neovim for having the separate languages, and also lifted some of their `scm` files (with attribution comments in this codebase) as the tree-sitter repo doesn't contain them. (Neovim's code is Apache-2.0 licensed, so should be fine here with attribution from reading Zed's licenses files.) I've then amended to make sure the capture groups are named for things Zed understands. I'd love someone from Zed to confirm that's okay, or if I should clean-room implement the `scm` files.

Highlighting in Terraform & HCL with a moderate amount of syntax in a file (Terraform on left, HCL on right.)

<img width="1392" alt="Screenshot 2024-01-31 at 18 07 45" src="https://github.com/zed-industries/zed/assets/696/1d3c9a08-588e-4b8f-ad92-98ce1e419659">

Release Notes:

- (|Improved) ... ([#5098](https://github.com/zed-industries/zed/issues/5098)).

caius committed:
Add terraform syntax highlighting

scm files mostly lifted from Neovim, under Apache-2.0 license

caius commented:
@cla-bot check

fdionisi commented:
@caius I have an initial implementation for the lsp on a branch - if you'd like we could pair and finish it off

SomeoneToIgnore commented:
This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

fdionisi commented:
Sent you an email already 😃

JosephTLyons commented:
Unless you've already started on the work, I think we could merge in the tree sitter portion of this and let the language server portion be a new PR.

caius commented:
> Unless you've already started on the work on this branch, I think we could merge in the tree sitter portion of this and let the language server portion be a new PR.

Good point @JosephTLyons, have marked this as ready for review to get the syntax highlighting shipped.

(I spy @fdionisi has :+1:'d you. We can open a second PR for shipping the LSP ✌🏻)

romans-ovo commented:
This branch uses default indent of 4, while terraform typically uses 2 as ident. Not sure if this is a relevant comment (my first one here).

caius committed:
Fix commenting in terraform

cmd-/ works for adding/removing comments

caius committed:
Default Terraform tab_size to 2 spaces

As per the HashiCorp Terraform [Style Conventions](https://developer.hashicorp.com/terraform/language/syntax/style)

caius commented on crates/zed/src/languages/terraform/config.toml:
```diff
@@ -0,0 +1,13 @@
+name = "Terraform"
```
I had this thought originally, given there's terraform-specific syntax in the highlighting and the LSP/Formatter is terraform-specific I figured it goes in as Terraform specifically.

I was then torn on adding Nomad, Packer separately or together as HCL as a second language albeit sharing the same tree-sitter-hcl package underneath. Those tend to be named `.nomad.hcl` or `.pkr.hcl` so being defined as HCL together makes sense I think.

fdionisi replied:
Indeed, but moving to hcl, would open another question 😅 can we trigger the LSP only on `.tf` and `.tfvars` using the same tree-sitter syntax? I'll try to dig on it between today and tomorrow, so we have all we need for our session

romans-ovo replied:
Terraform is a more specific language to HCL. HCL: https://developer.hashicorp.com/nomad/docs/job-specification/hcl2/syntax is more generic and is used across different tools from hashicorp and also 3rd party softawre such as Terragrunt.

Terraform files are `.tf` and HCL files are `.hcl`.

maxbrunsfeld replied:
In my opinion, we should just add a single language, and call it HCL. It's ok if the language server is Terraform specific and there is some Terraform-specific syntax highlighting.

caius replied:
We'll (well, I'll) want nomad formatting and packer formatting (which are basically the same I think) which differ to terraform formatting so I think the split in languages makes sense.

HCL is the base language for specifying configuration (and some dynamic parts) in, then Terraform extends that definition with terraform-specific functionality is my mental model of it. Hence why Packer & Nomad (and other tools) use `foo.hcl` named files and Terraform uses `foo.tf` named files.

I've pushed up a commit that adds HCL as a separate language.

caius replied:
Reading the terraform-ls docs[^1] are _very_ clear on only using it for `*.tf`, `*.tfvar` files, not sure Zed can be configured to only sends some filetypes it supports through to the LSP. Also a bit confusing to a user that the "HCL" language is chosen and their `*.tf` files are linted/error checked with terraform-ls, but if they open a `config.pkr.hcl` it's also using "HCL" language but not linted/error checked with the LSP.

I've not yet found a standalone hcl-lsp so I suspect we'll have to do without for Packer, Nomad files for now (ie, HCL) and add terraform-ls for Terraform files.

[^1]: https://github.com/hashicorp/terraform-ls/blob/main/docs/USAGE.md#usage-of-terraform-language-server

matkam replied:
@dbanck might be able to chime in here. he works at Hashicorp and offered to help: https://github.com/zed-industries/zed/issues/5098#issuecomment-1911727160

dbanck replied:
Thanks for the ping @matkam!

Looks like @caius and @fdionisi are on the right track. I think it makes sense to treat HCL and Terraform as different languages in Zed. This is how we do it for all other editors.

HCL is designed to be a toolkit for building languages, not a language itself, so it makes sense to provide only syntax highlighting via tree-stitter for it for now. There is some discussion about building a generic HCL formatter that applies basic formatting rules, but for now formatting is done by the applications that use HCL, such as `terraform fmt`, `packer fmt`, `vault policy fmt`, ...

For Terraform it makes sense to implement syntax highlighting via tree-sitter and language server support using https://github.com/hashicorp/terraform-ls. The language server can only handle Terraform-specific files (`.tf`, `.tf.json`, `.tfvars`, `.tfvars.json`, ...) and doesn't work well with generic `.hcl`. files. So it makes perfect sense to have separate language configurations.

There are no other language servers that I'm aware of, but if there is a Packer LS one day, it would need it's own language id and configuration (similar to Terraform today). This language configuration would claim the file extension `.pkr.hcl` and be more specific than the generic `.hcl` `path_suffixes`.

caius committed:
Add HCL as separate language

This is for Packer, Nomad, other infrastructure that uses HCL files, rather than terraform specific extensions.

dbanck commented on crates/zed/src/languages/hcl/config.toml:
```diff
@@ -0,0 +1,13 @@
+name = "HCL"
+path_suffixes = ["hcl", "hcl2"]
```
```suggestion
path_suffixes = ["hcl"]
```

I think it's safe to remove the `.hcl2` extension here. The [vscode-hcl extension doesn't claim it](https://github.com/hashicorp/vscode-hcl/blob/0ca34e6ce7dae6269858e39f687a2ac3732871ed/package.json#L47-L71) and it doesn't seem to be widely used.

caius replied:
Hah, old-er habits die hard. I can't remember the last time I encountered a hcl2 file to be fair. Good spot

caius committed:
Remove hcl2 extension from hcl language

caius commented on crates/zed/src/languages/hcl/config.toml:
```diff
@@ -0,0 +1,13 @@
+name = "HCL"
+path_suffixes = ["hcl", "hcl2"]
+line_comments = ["# "]
```
Hm, good point. I think I had `//` in there previously but backed it out for some reason. It's valid syntax so we should highlight it accordingly.

Seems the first element of this array is used by the cmd-/ when entering comments, so I think having it default to `#` makes sense even if it's different to vscode. 

Both Packer and Nomad default to `#` for comments in all the HCL examples. For Terraform, the [Syntax Guide](https://developer.hashicorp.com/terraform/language/syntax/configuration#comments) states

> The `#` single-line comment style is the default comment style and should be used in most cases. Automatic configuration formatting tools may automatically transform `//` comments into `#` comments, since the double-slash style is not idiomatic.

I think I'm okay with people opening code previously edited and being able to insert `//` by hand to continue their own convention, but using "editor: toggle comments" inserts `#` by default.

caius committed:
Allow correct line comments for Terraform & HCL

caius committed:
Define block_comment correctly for Terraform & HCL
